### ISSUE TITLE
Data Column Adjusted PCA

### ISSUE THREAD

**Issue by BrandonKMLee** [erdogant/pca#29]
Sometimes the base columnar data might not be normally distributed (bell curve) or continuously uniform (quantile-esque).

Guides on Power Transform (Yeo-Johnson vs Box-Cox):
- https://en.wikipedia.org/wiki/Power_transform
- https://statisticaloddsandends.wordpress.com/2021/02/19/the-box-cox-and-yeo-johnson-transformations-for-continuous-variables/
- https://jsmp.dk/posts/2019-08-23-transformingdata/

Library Choice: 
- https://feature-engine.readthedocs.io/en/1.0.x/transformation/YeoJohnsonTransformer.html
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson.html
- https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PowerTransformer.html

Demo on how scaling helps with PCA visualization https://towardsdatascience.com/feature-scaling-and-normalisation-in-a-nutshell-5319af86f89b
For non-transforms: Robust Scaling > Normalization > Standardization https://stats.stackexchange.com/questions/476394/impact-of-different-scaling-methods-on-pca-for-clustering

Q: is Yeo-Johnson good for quantile-esque columns?

Side note: this looks useful? https://github.com/erdogant/distfit

**Comment by BrandonKMLee** [comment]
Currently testing this to see which combinations makes a difference
```
from pandas import read_csv, merge
ace = save_unzip('https://files.catbox.moe/2nik4u.zip') # https://gist.github.com/regtm/5be7337561215c4e107b393312a17f2e
df = read_csv(ace + '/data-final.csv', sep='\t')
df = merge(df.iloc[:, : 50], df.iloc[:, -3:]).dropna()
X = df.iloc[:, : 50]
y = df.iloc[:, -3: ]

from pca import pca
from sklearn.preprocessing import RobustScaler, PowerTransformer, StandardScaler, MaxAbsScaler


fit_set = [RobustScaler, PowerTransformer, StandardScaler, MaxAbsScaler]

for i in fit_set:
  model = pca(n_components=0.8, verbose=2)
  results = model.fit_transform(i().fit_transform(X))
  model.plot(title=str(i))

from itertools import permutations 

for i in list(permutations(fit_set, 2)):
  model = pca(n_components=0.8, verbose=2)
  results = model.fit_transform(i[1]().fit_transform(i[0]().fit_transform(X)))
  model.plot(title=(str(i[0])+str(i[1])))
```

**Comment by BrandonKMLee** [comment]
Currently found out that PowerTransformer followed by RobustScaler (or just RobustScaler) performs the best compared to others [Curves.zip](https://github.com/erdogant/pca/files/9665022/Curves.zip)

Question:
1. How can these methods be compared against one another for usefulness/accuracy? Comparing against each other with a static N, or comparing N against minimum cumulative variance?
2. Would having automated feature engineering produce similar speedups? If so, how would one benchmark against different tables with different column counts? https://github.com/feature-engine/feature_engine

**Comment by BrandonKMLee** [comment]
For reference, this is the personality dataset, and these are the 50 questions regarding oneself ("Likehart Scale" that tends to follow a skewed bell curve). Some of the items should be strongly correlated but the inter-cluster correlations may be fuzzy. The original idea is to reduce this into clusters of "major personalities" rather than "personality meta-vectors".

Also nice blog on the comparison, having [PHATE](https://twitter.com/KrishnaswamyLab/status/1201935823056199680), [TriMap, PaCMAP,](https://www.kaggle.com/code/frankmollard/a-story-about-unsupervised-learning?scriptVersionId=80178686), [LargeVis](https://www.medrxiv.org/content/10.1101/2021.04.01.21254751v1.full), [TOPAOE, ATSNE, UMATO](https://openreview.net/pdf?id=gkOYZpeGEK), and others would be a good primer on local preservation vs global preservation.

**Comment by erdogant** [comment]
ok! This[one](https://www.kaggle.com/datasets/tunguz/big-five-personality-test) ?

**Comment by BrandonKMLee** [comment]
Yes, and one thing special about this, is that most survey data:
- are somewhat correlated in facet level (correlation between dimensional cluster aggregates exists e.g. [Plasticity and Stability](https://www.kaggle.com/code/psqrtpsqrt/big-five-personality-test-eda/notebook))
- can often be correlated in the item level (correlation between questions of different clusters exists e.g. [Correlational Heat Map](https://www.kaggle.com/code/zeynepsivri/five-personality-clustering-k-means-dbscan/notebook))
- are difficult to cluster in regards to data points ([t-SNE attempt](https://www.kaggle.com/code/khotijahs1/big-five-personality-test-clustering) and [YellowBrick](https://www.kaggle.com/code/gokceergun/cluster-analysis-five-factor-personality-model/notebook))
- each column have discrete value instead of continuous, which makes normalization and scaling difficult

If one wants to try these techniques on other similar datasets see OpenPsychometrics in http://openpsychometrics.org/_rawdata/

**Comment by erdogant** [comment]
closing this one.
